File: host/signer_host/procs.py
"""Programs that can hold a token's single PKCS#11 session.

ePass/ProxKey-class drivers allow one process on the token at a time; a
vendor utility, another browser's host, or a competing signing host silently makes
every scan come back empty. Naming the culprit beats "replug and retry".
Tolerant like pcsc.py: any failure means "nothing found", never an error.
"""

import os
import subprocess
import sys

# lowercase needle in the process name -> what to tell the user to close.
_KNOWN = (
    ("opensigner-host", "another OpenSigner host"),
    ("webpki", "a competing signing host"),
    ("epass", "the ePass token manager"),
    ("proxkey", "the ProxKey token tool"),
    ("wdtoken", "the WatchData token tool"),
    ("etoken", "SafeNet Authentication Client"),
    ("safenet", "SafeNet Authentication Client"),
    ("cryptoida", "the mToken CryptoID tool"),
    ("trustkey", "the TrustKey tool"),
)


def _process_list():
    """(pid, lowercased process name) for every visible process; [] on failure."""
    try:
        if sys.platform == "win32":
            out = subprocess.run(
                ["tasklist", "/fo", "csv", "/nh"],
                capture_output=True, text=True, timeout=10,
            ).stdout
            rows = []
            for line in out.splitlines():
                parts = [p.strip('"') for p in line.split('","')]
                if len(parts) >= 2 and parts[1].isdigit():
                    rows.append((int(parts[1]), parts[0].lower()))
            return rows
        out = subprocess.run(
            ["ps", "-A", "-o", "pid=,comm="],
            capture_output=True, text=True, timeout=10,
        ).stdout
        rows = []
        for line in out.splitlines():
            pid, _, name = line.strip().partition(" ")
            if pid.isdigit() and name:
                rows.append((int(pid), os.path.basename(name.strip()).lower()))
        return rows
    except Exception:
        return []


def competing():
    """Display names of running programs likely holding the token, deduplicated.

    Our own process tree is excluded. ponytail: self = pid + ppid, which
    covers PyInstaller's one-file parent/child pair; a third opensigner-host
    of our own tree would be a real finding anyway.
    """
    own = {os.getpid(), os.getppid()}
    found = []
    for pid, name in _process_list():
        if pid in own:
            continue
        for needle, label in _KNOWN:
            if needle in name and label not in found:
                found.append(label)
    return found

File: host-rs/src/procs.rs
//! Programs that can hold a token's single PKCS#11 session.
//!
//! ePass/ProxKey-class drivers allow one process on the token at a time; a
//! vendor utility, another browser's host, or a third-party signing host
//! silently makes every scan come back empty. Naming the culprit beats
//! "replug and retry".
//! Tolerant like `pcsc_readers`: any failure means "nothing found", never an error.
//!
//! ponytail: shells out to ps/tasklist rather than pulling sysinfo, which is
//! ~500 KB of crate for one process list on a path that only runs when a scan
//! already came back empty.

use std::process::Command;

/// Lowercase needle in the process name -> what to tell the user to close.
const KNOWN: &[(&str, &str)] = &[
    ("docsigner-host", "another DocSigner host"),
    ("webpki", "a Web PKI signing host"),
    ("epass", "the ePass token manager"),
    ("proxkey", "the ProxKey token tool"),
    ("wdtoken", "the WatchData token tool"),
    ("etoken", "SafeNet Authentication Client"),
    ("safenet", "SafeNet Authentication Client"),
    ("cryptoida", "the mToken CryptoID tool"),
    ("trustkey", "the TrustKey tool"),
];

/// Display names of running programs likely holding the token, deduplicated.
///
/// Our own process is excluded. ponytail: self = this pid only; Rust has no
/// PyInstaller parent/child pair to filter, unlike the Python host.
pub fn competing() -> Vec<String> {
    let own = std::process::id();
    let mut found: Vec<String> = Vec::new();
    for (pid, name) in process_list() {
        if pid == own {
            continue;
        }
        for (needle, label) in KNOWN {
            if name.contains(needle) && !found.iter().any(|f| f == label) {
                found.push((*label).to_string());
            }
        }
    }
    found
}

/// (pid, lowercased process name) for every visible process; empty on failure.
fn process_list() -> Vec<(u32, String)> {
    #[cfg(target_os = "windows")]
    let output = Command::new("tasklist")
        .args(["/fo", "csv", "/nh"])
        .output();
    #[cfg(not(target_os = "windows"))]
    let output = Command::new("ps").args(["-A", "-o", "pid=,comm="]).output();

    let Ok(output) = output else {
        return Vec::new();
    };
    let text = String::from_utf8_lossy(&output.stdout);

    #[cfg(target_os = "windows")]
    {
        parse_tasklist(&text)
    }
    #[cfg(not(target_os = "windows"))]
    {
        parse_ps(&text)
    }
}

#[cfg(not(target_os = "windows"))]
fn parse_ps(text: &str) -> Vec<(u32, String)> {
    let mut rows = Vec::new();
    for line in text.lines() {
        let line = line.trim();
        let Some((pid, name)) = line.split_once(' ') else {
            continue;
        };
        let Ok(pid) = pid.trim().parse::<u32>() else {
            continue;
        };
        let name = name.trim();
        if name.is_empty() {
            continue;
        }
        let base = name.rsplit('/').next().unwrap_or(name);
        rows.push((pid, base.to_lowercase()));
    }
    rows
}

#[cfg(target_os = "windows")]
fn parse_tasklist(text: &str) -> Vec<(u32, String)> {
    let mut rows = Vec::new();
    for line in text.lines() {
        // "name.exe","1234","Console","1","10,000 K"
        let parts: Vec<&str> = line.split("\",\"").collect();
        if parts.len() < 2 {
            continue;
        }
        let name = parts[0].trim_matches('"');
        let Ok(pid) = parts[1].trim_matches('"').parse::<u32>() else {
            continue;
        };
        rows.push((pid, name.to_lowercase()));
    }
    rows
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn competing_never_panics() {
        let _ = competing();
    }

    /// Runs the real ps/tasklist and checks the parse, since the format is the
    /// only thing likely to drift.
    ///
    /// Deliberately no assertion that every pid is non-zero: Windows lists the
    /// System Idle Process as pid 0, which is a real row, not a parse failure.
    #[test]
    fn the_real_process_list_parses() {
        let list = process_list();
        assert!(!list.is_empty(), "expected at least one visible process");
        assert!(
            list.iter().all(|(_, name)| !name.is_empty()),
            "a parsed row must carry a name, or the needle match below is dead"
        );
        assert!(
            list.iter().all(|(_, name)| name == &name.to_lowercase()),
            "names are lowercased so the KNOWN needles can match"
        );
        // Our own process must be in there, which is what `competing` filters on.
        let own = std::process::id();
        assert!(
            list.iter().any(|(pid, _)| *pid == own),
            "the running test process should appear in its own process list"
        );
    }

    #[cfg(not(target_os = "windows"))]
    #[test]
    fn parses_ps_output() {
        let sample = "  501 /usr/sbin/cfprefsd\n  777 WatchData-ProxKey\n bad line\n";
        let rows = parse_ps(sample);
        assert_eq!(rows.len(), 2);
        assert_eq!(rows[0], (501, "cfprefsd".to_string()));
        assert_eq!(rows[1], (777, "watchdata-proxkey".to_string()));
    }

    #[cfg(target_os = "windows")]
    #[test]
    fn parses_tasklist_output() {
        let sample = "\"eToken.exe\",\"1234\",\"Console\",\"1\",\"10,000 K\"\n";
        let rows = parse_tasklist(sample);
        assert_eq!(rows, vec![(1234, "etoken.exe".to_string())]);
    }

    #[test]
    fn known_needles_are_lowercase() {
        for (needle, label) in KNOWN {
            assert_eq!(*needle, needle.to_lowercase());
            assert!(!label.is_empty());
        }
    }
}
